File: backend/protocol/voting.py
"""投票相关协议类型"""

from dataclasses import dataclass
from enum import Enum


class Stance(str, Enum):
    SUPPORT = "support"
    OPPOSE = "oppose"
    MODIFY = "modify"
    NEUTRAL = "neutral"


class ConsensusStrategy(str, Enum):
    SIMPLE_MAJORITY = "simple_majority"


@dataclass
class Proposal:
    id: str
    proposer_id: str
    content: str
    stance: Stance = Stance.NEUTRAL
    confidence: float = 1.0
    created_at: float = 0.0


@dataclass
class Vote:
    proposal_id: str
    voter_id: str
    approve: bool
    weight: float = 1.0
    reason: str = ""


@dataclass
class VoteResult:
    proposal_id: str
    total_votes: int = 0
    approve_count: int = 0
    oppose_count: int = 0
    accepted: bool = False


# ── 序列化函数 ──

def proposal_to_dict(proposal: Proposal) -> dict:
    return {
        "id": proposal.id,
        "proposer_id": proposal.proposer_id,
        "content": proposal.content,
        "stance": proposal.stance.value,
        "confidence": proposal.confidence,
        "created_at": proposal.created_at,
    }


def vote_to_dict(vote: Vote) -> dict:
    return {
        "proposal_id": vote.proposal_id,
        "voter_id": vote.voter_id,
        "approve": vote.approve,
        "weight": vote.weight,
        "reason": vote.reason,
    }


def vote_result_to_dict(result: VoteResult) -> dict:
    return {
        "proposal_id": result.proposal_id,
        "total_votes": result.total_votes,
        "approve_count": result.approve_count,
        "oppose_count": result.oppose_count,
        "accepted": result.accepted,
    }


def dict_to_proposal(data: dict) -> Proposal:
    return Proposal(
        id=data["id"],
        proposer_id=data["proposer_id"],
        content=data["content"],
        stance=Stance(data["stance"]) if "stance" in data else Stance.NEUTRAL,
        confidence=data.get("confidence", 1.0),
        created_at=data.get("created_at", 0.0),
    )


def dict_to_vote(data: dict) -> Vote:
    return Vote(
        proposal_id=data["proposal_id"],
        voter_id=data["voter_id"],
        approve=data["approve"],
        weight=data.get("weight", 1.0),
        reason=data.get("reason", ""),
    )


def dict_to_vote_result(data: dict) -> VoteResult:
    return VoteResult(
        proposal_id=data["proposal_id"],
        total_votes=data.get("total_votes", 0),
        approve_count=data.get("approve_count", 0),
        oppose_count=data.get("oppose_count", 0),
        accepted=data.get("accepted", False),
    )

File: src/hooks/useMeetingSocket/handlers/voting.ts
/**
 * 投票相关消息处理器
 */

import type { ChatMessage } from '../../components/office-team/types'

export interface VoteMessage {
  proposal?: { id: string; proposerId: string; content: string; createdAt: string }
  vote?: { voterId: string; approve: boolean; reason?: string }
  result?: { proposalId: string; totalVotes: number; approveCount: number; opposeCount: number; accepted: boolean }
}

export interface VotingSetters {
  setChatMessages: (fn: (prev: ChatMessage[]) => ChatMessage[]) => void
  setActiveProposal: (p: { id: string; proposerId: string; content: string; createdAt: string } | null) => void
  setVotes: (fn: (prev: Map<string, { voterId: string; approve: boolean; reason?: string }>) => Map<string, { voterId: string; approve: boolean; reason?: string }>) => void
  setVoteResults: (r: { proposalId: string; totalVotes: number; approveCount: number; opposeCount: number; accepted: boolean } | null) => void
}

export function handleProposal(msg: VoteMessage, setters: VotingSetters) {
  const proposal = msg.proposal
  if (proposal) {
    setters.setActiveProposal({
      id: proposal.id,
      proposerId: proposal.proposerId,
      content: proposal.content,
      createdAt: proposal.createdAt,
    })
    setters.setVotes(() => new Map())
    setters.setVoteResults(null)
    setters.setChatMessages(prev => [...prev, {
      role: 'agent' as const,
      agentId: proposal.proposerId,
      content: `[提案] ${proposal.content}`,
      timestamp: Date.now(),
      _msgSubtype: 'proposal',
    }])
  }
}

export function handleVote(msg: VoteMessage, setters: VotingSetters) {
  const vote = msg.vote
  if (vote) {
    setters.setVotes(prev => {
      const next = new Map(prev)
      next.set(vote.voterId, {
        voterId: vote.voterId,
        approve: vote.approve,
        reason: vote.reason,
      })
      return next
    })
    setters.setChatMessages(prev => [...prev, {
      role: 'agent' as const,
      agentId: vote.voterId,
      content: `[投票] ${vote.approve ? '赞成' : '反对'}${vote.reason ? ': ' + vote.reason : ''}`,
      timestamp: Date.now(),
      _msgSubtype: 'vote',
    }])
  }
}

export function handleVoteResult(msg: VoteMessage, setters: VotingSetters) {
  const result = msg.result
  if (result) {
    setters.setVoteResults({
      proposalId: result.proposalId,
      totalVotes: result.totalVotes,
      approveCount: result.approveCount,
      opposeCount: result.opposeCount,
      accepted: result.accepted,
    })
    setters.setActiveProposal(null)
    setters.setVotes(() => new Map())
    setters.setChatMessages(prev => [...prev, {
      role: 'boss' as const,
      content: `投票结果: ${result.accepted ? '通过' : '未通过'} (${result.approveCount}赞成 / ${result.opposeCount}反对，共${result.totalVotes}票)`,
      timestamp: Date.now(),
      _msgSubtype: 'vote_result',
    }])
  }
}
